File: .agents/cli/core/decomposition_analysis.py
"""Read-only parent/child decomposition coverage helpers for q-analyze."""
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any, Iterable, Mapping

import yaml
from jsonschema import Draft202012Validator

LOCATIONS = ("inbox", "active", "done", "failed")
PARENT_ID_RE = re.compile(r"^[A-Z]+-[0-9]+$")


def analyze_parent_child_coverage(parent_spec_path, ai_tasks_root, spec_schema_path=None):
    """Return coverage/findings for a decomposed parent without mutating state."""
    parent_spec_path = Path(parent_spec_path)
    ai_tasks_root = Path(ai_tasks_root)
    schema = _schema(spec_schema_path)
    parent, error = _load_spec(parent_spec_path, schema)
    if error:
        finding = _finding("critical", parent_spec_path.as_posix(), f"Parent spec could not be loaded: {error}")
        return _result(False, "blocked", findings=[finding])

    parent_id = str(parent.get("task_id") or parent_spec_path.parent.name)
    decomposition = parent.get("decomposition") or []
    if not decomposition:
        out = _result(False, "not_decomposed")
        out.update(parent_task_id=parent_id, parent_spec_path=parent_spec_path.as_posix())
        return out
    if not isinstance(decomposition, list):
        finding = _finding("critical", "00-spec.yaml.decomposition", "Expected decomposition to be a list.")
        out = _result(True, "blocked", findings=[finding])
        out.update(parent_task_id=parent_id, parent_spec_path=parent_spec_path.as_posix())
        return out

    entries, findings = _entries(decomposition)
    declared = {entry["child_id"] for entry in entries}
    children: dict[str, dict[str, Any]] = {}
    loaded: dict[str, Mapping[str, Any]] = {}

    for entry in entries:
        child_id = entry["child_id"]
        expected = list(entry.get("depends_on") or [])
        path, resolver_error = _find_spec_path(child_id, ai_tasks_root)
        if resolver_error or path is None:
            issue = resolver_error or f"Child {child_id} is declared but no 00-spec.yaml was found."
            children[child_id] = {"status": "missing", "path": None, "error": issue, "depends_on": expected}
            findings.append(_finding("high", f"00-spec.yaml.decomposition[{child_id}]", issue))
            continue

        spec, error = _load_spec(path, schema)
        children[child_id] = {"status": "invalid" if error else "loaded", "path": path.as_posix(), "error": error, "depends_on": expected}
        if error:
            findings.append(_finding("high", path.as_posix(), f"Child {child_id} has an invalid 00-spec.yaml: {error}"))
            continue

        loaded[child_id] = spec
        if spec.get("parent_task") != parent_id:
            findings.append(_finding("high", f"{path.as_posix()}.parent_task", f"Child {child_id} references parent_task={spec.get('parent_task')!r}; expected {parent_id!r}."))
        actual = list(spec.get("depends_on") or [])
        unknown = sorted(dep for dep in actual if dep not in declared)
        if unknown:
            findings.append(_finding("medium", f"{path.as_posix()}.depends_on", f"Child {child_id} depends on undeclared sibling(s): {', '.join(unknown)}."))
        if set(actual) != set(expected):
            findings.append(_finding("medium", f"{path.as_posix()}.depends_on", f"Child {child_id} depends_on {sorted(actual)} does not match parent decomposition {sorted(expected)}."))

    for entry in entries:
        for dep in entry.get("depends_on") or []:
            if dep not in declared:
                findings.append(_finding("medium", f"00-spec.yaml.decomposition[{entry['child_id']}].depends_on", f"Child {entry['child_id']} depends on undeclared sibling {dep}."))

    coverage = {
        "invariants": _coverage_for_items(parent.get("invariants") or [], loaded, "invariants"),
        "acceptance": _coverage_for_items(parent.get("acceptance") or [], loaded, "acceptance"),
    }
    gaps = _gaps(coverage)
    findings.extend(gaps)
    return {
        "applies": True,
        "status": "issues_found" if findings else "pass",
        "parent_task_id": parent_id,
        "parent_spec_path": parent_spec_path.as_posix(),
        "children": children,
        "coverage": coverage,
        "gaps": gaps,
        "inconsistencies": [finding for finding in findings if finding not in gaps],
        "findings": findings,
    }


def load_parent_child_specs(parent_spec_path, ai_tasks_root, spec_schema_path=None):
    """Return parent/child load state without coverage rows."""
    result = analyze_parent_child_coverage(parent_spec_path, ai_tasks_root, spec_schema_path)
    return {key: result.get(key) for key in ("applies", "status", "parent_task_id", "parent_spec_path", "children", "inconsistencies")}


def _coverage_for_items(parent_items: Iterable[str], child_specs: Mapping[str, Mapping[str, Any]], field: str):
    rows = []
    for item in map(str, parent_items):
        covered_by = [child_id for child_id, spec in sorted(child_specs.items()) if any(_covers(item, str(child_item)) for child_item in spec.get(field) or [])]
        rows.append({"item": item, "covered_by": covered_by})
    return rows


def _gaps(coverage):
    gaps = []
    for field, rows in coverage.items():
        for index, row in enumerate(rows):
            if not row["covered_by"]:
                name = "invariant" if field == "invariants" else "acceptance"
                severity = "medium" if field == "invariants" else "high"
                gaps.append(_finding(severity, f"00-spec.yaml.{field}[{index}]", f"No child spec covers parent {name}: {row['item']}"))
    return gaps


def _entries(decomposition):
    entries, findings, seen = [], [], set()
    for index, entry in enumerate(decomposition):
        artifact = f"00-spec.yaml.decomposition[{index}]"
        if not isinstance(entry, Mapping):
            findings.append(_finding("high", artifact, "Expected decomposition entry to be an object.")); continue
        child_id = entry.get("child_id")
        if not child_id:
            findings.append(_finding("high", artifact, "Missing child_id.")); continue
        child_id = str(child_id)
        if child_id in seen:
            findings.append(_finding("high", artifact, f"Duplicate child_id {child_id}.")); continue
        seen.add(child_id)
        entries.append({"child_id": child_id, "depends_on": [str(dep) for dep in entry.get("depends_on") or []]})
    return entries, findings


def _find_spec_path(task_id: str, ai_tasks_root: Path):
    yaml_matches, exact_matches, prefix_matches = [], [], []
    for location in LOCATIONS:
        root = ai_tasks_root / location
        if not root.exists():
            continue
        for task_dir in sorted(path for path in root.iterdir() if path.is_dir()):
            spec_path = task_dir / "00-spec.yaml"
            spec_id = _read_task_id(spec_path)
            if spec_id == task_id:
                yaml_matches.append(spec_path); continue
            if task_dir.name == task_id:
                exact_matches.append(spec_path); continue
            if task_dir.name.startswith(f"{task_id}-"):
                rest = task_dir.name[len(task_id) + 1:]
                if not (PARENT_ID_RE.match(task_id) and _child_suffix(rest)):
                    prefix_matches.append(spec_path)
    matches = yaml_matches or exact_matches or prefix_matches
    if len(matches) > 1:
        return None, f"Multiple task directories match {task_id}: {', '.join(path.parent.as_posix() for path in matches)}."
    return (matches[0], None) if matches else (None, None)


def _read_task_id(spec_path: Path):
    if not spec_path.exists():
        return None
    try:
        spec = yaml.safe_load(spec_path.read_text())
    except (OSError, yaml.YAMLError):
        return None
    return str(spec.get("task_id")) if isinstance(spec, Mapping) and spec.get("task_id") else None


def _load_spec(path: Path, schema):
    try:
        spec = yaml.safe_load(path.read_text())
    except OSError as error:
        return None, str(error)
    except yaml.YAMLError as error:
        return None, str(error)
    if not isinstance(spec, Mapping):
        return None, "spec root is not an object"
    if schema is not None:
        errors = sorted(Draft202012Validator(schema).iter_errors(spec), key=lambda err: list(err.path))
        if errors:
            first = errors[0]
            field = ".".join(str(part) for part in first.path) or "$"
            return spec, f"field={field}; reason={first.message}"
    return spec, None


def _schema(path):
    if path is None:
        return None
    try:
        return json.loads(Path(path).read_text())
    except (OSError, ValueError):
        return None


def _covers(parent_item: str, child_item: str):
    parent_norm = _norm(parent_item)
    child_norm = _norm(child_item)
    return bool(parent_norm) and (parent_norm == child_norm or parent_norm in child_norm)


def _norm(value: str):
    return re.sub(r"\W+", " ", value.lower()).strip()


def _child_suffix(rest: str):
    return (len(rest) == 1 and rest.islower()) or (len(rest) > 1 and rest[0].islower() and rest[1] == "-")


def _finding(severity: str, artifact: str, issue: str):
    return {"severity": severity, "artifact": artifact, "issue": issue}


def _result(applies: bool, status: str, findings=None):
    findings = list(findings or [])
    return {"applies": applies, "status": status, "parent_task_id": None, "parent_spec_path": None, "children": {}, "coverage": {"invariants": [], "acceptance": []}, "gaps": [], "inconsistencies": findings, "findings": findings}

File: internal/core/decomposition_analysis.go
package core

import (
	"fmt"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

type Finding struct {
	Severity string `json:"severity"`
	Artifact string `json:"artifact"`
	Issue    string `json:"issue"`
}

type CoverageRow struct {
	Item      string   `json:"item"`
	CoveredBy []string `json:"covered_by"`
}

type Coverage struct {
	Invariants []CoverageRow `json:"invariants"`
	Acceptance []CoverageRow `json:"acceptance"`
}

type ChildStatus struct {
	Status    string   `json:"status"`
	Path      *string  `json:"path"`
	Error     *string  `json:"error"`
	DependsOn []string `json:"depends_on"`
}

type DecompositionAnalysisResult struct {
	Applies         bool                   `json:"applies"`
	Status          string                 `json:"status"`
	ParentTaskID    *string                `json:"parent_task_id"`
	ParentSpecPath  *string                `json:"parent_spec_path"`
	Children        map[string]ChildStatus `json:"children"`
	Coverage        Coverage               `json:"coverage"`
	Gaps            []Finding              `json:"gaps"`
	Inconsistencies []Finding              `json:"inconsistencies"`
	Findings        []Finding              `json:"findings"`
}

func AnalyzeParentChildCoverage(parentSpecPath, aiTasksRoot string) DecompositionAnalysisResult {
	res := DecompositionAnalysisResult{
		Applies:         false,
		Status:          "not_decomposed",
		Children:        make(map[string]ChildStatus),
		Gaps:            []Finding{},
		Inconsistencies: []Finding{},
		Findings:        []Finding{},
		Coverage: Coverage{
			Invariants: []CoverageRow{},
			Acceptance: []CoverageRow{},
		},
	}

	payload, err := LoadArtifactPayload(parentSpecPath)
	if err != nil {
		res.Status = "blocked"
		msg := "Parent spec could not be loaded: " + err.Error()
		res.Findings = append(res.Findings, Finding{"critical", filepath.ToSlash(parentSpecPath), msg})
		return res
	}

	parentSpec, ok := payload.(map[string]any)
	if !ok {
		res.Status = "blocked"
		msg := "Parent spec could not be loaded: spec root is not an object"
		res.Findings = append(res.Findings, Finding{"critical", filepath.ToSlash(parentSpecPath), msg})
		return res
	}

	// Validate parent (using core.ValidateArtifact)
	if err := ValidateArtifact(parentSpecPath, parentSpec); err != nil {
		var reason string
		if ve, ok := err.(ArtifactValidationError); ok {
			reason = ve.Message
		} else {
			reason = err.Error()
		}
		res.Status = "blocked"
		msg := "Parent spec could not be loaded: " + reason
		res.Findings = append(res.Findings, Finding{"critical", filepath.ToSlash(parentSpecPath), msg})
		return res
	}

	taskID := ""
	if id, ok := parentSpec["task_id"].(string); ok && id != "" {
		taskID = id
	} else {
		taskID = filepath.Base(filepath.Dir(parentSpecPath))
	}
	res.ParentTaskID = &taskID
	psp := filepath.ToSlash(parentSpecPath)
	res.ParentSpecPath = &psp

	decompObj := parentSpec["decomposition"]
	if decompObj == nil {
		return res
	}

	decompList, ok := asSlice(decompObj)
	if !ok {
		res.Applies = true
		res.Status = "blocked"
		res.Findings = append(res.Findings, Finding{"critical", "00-spec.yaml.decomposition", "Expected decomposition to be a list."})
		return res
	}

	res.Applies = true

	var entries []map[string]any
	var findings []Finding
	seen := make(map[string]bool)

	for i, entryAny := range decompList {
		artifact := fmt.Sprintf("00-spec.yaml.decomposition[%d]", i)
		entry, ok := entryAny.(map[string]any)
		if !ok {
			findings = append(findings, Finding{"high", artifact, "Expected decomposition entry to be an object."})
			continue
		}
		childID, ok := entry["child_id"].(string)
		if !ok || childID == "" {
			findings = append(findings, Finding{"high", artifact, "Missing child_id."})
			continue
		}
		if seen[childID] {
			findings = append(findings, Finding{"high", artifact, "Duplicate child_id " + childID + "."})
			continue
		}
		seen[childID] = true
		entries = append(entries, entry)
	}

	loadedChildren := make(map[string]map[string]any)

	for _, entry := range entries {
		childID := entry["child_id"].(string)
		var expected []string
		if depsAny, ok := asSlice(entry["depends_on"]); ok {
			for _, d := range depsAny {
				if ds, ok := d.(string); ok {
					expected = append(expected, ds)
				}
			}
		}

		childDirMatch, err := FindTaskDirIn(filepath.Dir(filepath.Dir(aiTasksRoot)), childID, []string{"inbox", "active", "done", "failed"})
		var childSpecPath string
		var resolverError string

		if err != nil {
			resolverError = err.Error()
		} else if childDirMatch == nil {
			resolverError = "Child " + childID + " is declared but no 00-spec.yaml was found."
		} else {
			childSpecPath = filepath.Join(childDirMatch.Path, "00-spec.yaml")
		}

		if resolverError != "" {
			res.Children[childID] = ChildStatus{
				Status:    "missing",
				Error:     &resolverError,
				DependsOn: expected,
			}
			findings = append(findings, Finding{"high", "00-spec.yaml.decomposition[" + childID + "]", resolverError})
			continue
		}

		csp := filepath.ToSlash(childSpecPath)
		childPayload, err := LoadArtifactPayload(childSpecPath)
		var loadError string
		if err != nil {
			loadError = err.Error()
		} else {
			if _, ok := childPayload.(map[string]any); !ok {
				loadError = "spec root is not an object"
			} else {
				if err := ValidateArtifact(childSpecPath, childPayload); err != nil {
					if ve, ok := err.(ArtifactValidationError); ok {
						loadError = ve.Message
					} else {
						loadError = err.Error()
					}
				}
			}
		}

		if loadError != "" {
			res.Children[childID] = ChildStatus{
				Status:    "invalid",
				Path:      &csp,
				Error:     &loadError,
				DependsOn: expected,
			}
			findings = append(findings, Finding{"high", csp, "Child " + childID + " has an invalid 00-spec.yaml: " + loadError})
			continue
		}

		childSpec := childPayload.(map[string]any)
		res.Children[childID] = ChildStatus{
			Status:    "loaded",
			Path:      &csp,
			DependsOn: expected,
		}
		loadedChildren[childID] = childSpec

		if pt, ok := childSpec["parent_task"].(string); !ok || pt != taskID {
			actualParent := ""
			if ok {
				actualParent = pt
			}
			findings = append(findings, Finding{"high", csp + ".parent_task", "Child " + childID + " references parent_task='" + actualParent + "'; expected '" + taskID + "'."})
		}

		var actual []string
		if depsAny, ok := asSlice(childSpec["depends_on"]); ok {
			for _, d := range depsAny {
				if ds, ok := d.(string); ok {
					actual = append(actual, ds)
				}
			}
		}

		var unknown []string
		for _, a := range actual {
			if !seen[a] {
				unknown = append(unknown, a)
			}
		}
		if len(unknown) > 0 {
			sort.Strings(unknown)
			findings = append(findings, Finding{"medium", csp + ".depends_on", "Child " + childID + " depends on undeclared sibling(s): " + strings.Join(unknown, ", ") + "."})
		}

		if !stringSlicesEqual(actual, expected) {
			aCopy := append([]string(nil), actual...)
			eCopy := append([]string(nil), expected...)
			sort.Strings(aCopy)
			sort.Strings(eCopy)
			aStr := "[]"
			if len(aCopy) > 0 {
				aStr = "['" + strings.Join(aCopy, "', '") + "']"
			}
			eStr := "[]"
			if len(eCopy) > 0 {
				eStr = "['" + strings.Join(eCopy, "', '") + "']"
			}
			findings = append(findings, Finding{"medium", csp + ".depends_on", "Child " + childID + " depends_on " + aStr + " does not match parent decomposition " + eStr + "."})
		}
	}

	for _, entry := range entries {
		childID := entry["child_id"].(string)
		if depsAny, ok := asSlice(entry["depends_on"]); ok {
			for _, d := range depsAny {
				if ds, ok := d.(string); ok && !seen[ds] {
					findings = append(findings, Finding{"medium", "00-spec.yaml.decomposition[" + childID + "].depends_on", "Child " + childID + " depends on undeclared sibling " + ds + "."})
				}
			}
		}
	}

	res.Coverage.Invariants = coverageForItems(parentSpec["invariants"], loadedChildren, "invariants")
	res.Coverage.Acceptance = coverageForItems(parentSpec["acceptance"], loadedChildren, "acceptance")

	for i, row := range res.Coverage.Invariants {
		if len(row.CoveredBy) == 0 {
			findings = append(findings, Finding{"medium", fmt.Sprintf("00-spec.yaml.invariants[%d]", i), "No child spec covers parent invariant: " + row.Item})
			res.Gaps = append(res.Gaps, findings[len(findings)-1])
		}
	}
	for i, row := range res.Coverage.Acceptance {
		if len(row.CoveredBy) == 0 {
			findings = append(findings, Finding{"high", fmt.Sprintf("00-spec.yaml.acceptance[%d]", i), "No child spec covers parent acceptance: " + row.Item})
			res.Gaps = append(res.Gaps, findings[len(findings)-1])
		}
	}

	res.Findings = findings
	for _, f := range findings {
		isGap := false
		for _, g := range res.Gaps {
			if f.Issue == g.Issue && f.Artifact == g.Artifact {
				isGap = true
				break
			}
		}
		if !isGap {
			res.Inconsistencies = append(res.Inconsistencies, f)
		}
	}

	if len(findings) > 0 {
		res.Status = "issues_found"
	} else {
		res.Status = "pass"
	}

	return res
}

func coverageForItems(parentItemsAny any, loadedChildren map[string]map[string]any, field string) []CoverageRow {
	var rows []CoverageRow
	parentItems, ok := asSlice(parentItemsAny)
	if !ok {
		return rows
	}

	childIDs := []string{}
	for id := range loadedChildren {
		childIDs = append(childIDs, id)
	}
	sort.Strings(childIDs)

	for _, piAny := range parentItems {
		pi := acceptanceStatement(piAny)
		var coveredBy []string
		for _, childID := range childIDs {
			childSpec := loadedChildren[childID]
			childItemsAny := childSpec[field]
			if childItems, ok := asSlice(childItemsAny); ok {
				covers := false
				for _, ciAny := range childItems {
					ci := acceptanceStatement(ciAny)
					if coversItem(pi, ci) {
						covers = true
						break
					}
				}
				if covers {
					coveredBy = append(coveredBy, childID)
				}
			}
		}
		rows = append(rows, CoverageRow{Item: pi, CoveredBy: coveredBy})
	}
	return rows
}

func coversItem(parentItem, childItem string) bool {
	pNorm := normStr(parentItem)
	cNorm := normStr(childItem)
	return pNorm != "" && (pNorm == cNorm || strings.Contains(cNorm, pNorm))
}

var nonWordRe = regexp.MustCompile(`\W+`)

func normStr(s string) string {
	return strings.TrimSpace(nonWordRe.ReplaceAllString(strings.ToLower(s), " "))
}

func stringSlicesEqual(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	aMap := make(map[string]bool)
	for _, s := range a {
		aMap[s] = true
	}
	for _, s := range b {
		if !aMap[s] {
			return false
		}
	}
	return true
}
